SonarQube.Client/Models/ServerSentEvents/SSEStreamWriter.cs

```diff
@@ -35,39 +36,60 @@ public interface ISSEStreamWriter : IDisposable
         Task BeginListening();
     }
 
+    /// <summary>
+    /// Aggregates stream lines into events.
+    /// Code on the java side: https://github.com/SonarSource/sonarlint-core/blob/171ca4d75c24033e115a81bd7481427cd1f39f4c/server-api/src/main/java/org/sonarsource/sonarlint/core/serverapi/stream/EventBuffer.java
+    /// </summary>
     internal class SSEStreamWriter : ISSEStreamWriter
     {
-        private readonly Stream serverStream;
+        private readonly StreamReader networkStreamReader;
         private readonly ChannelWriter<ISqServerEvent> sqEventsChannel;
         private readonly CancellationToken cancellationToken;
         private readonly ISqServerSentEventParser sqServerSentEventParser;
 
-        public SSEStreamWriter(Stream serverStream,
+        public SSEStreamWriter(StreamReader networkStreamReader,
             ChannelWriter<ISqServerEvent> sqEventsChannel,
             CancellationToken cancellationToken)
-            : this(serverStream, sqEventsChannel, cancellationToken, new SqServerSentEventParser())
+            : this(networkStreamReader, sqEventsChannel, cancellationToken, new SqServerSentEventParser())
         {
         }
 
-        internal SSEStreamWriter(Stream serverStream,
+        internal SSEStreamWriter(StreamReader networkStreamReader,
             ChannelWriter<ISqServerEvent> sqEventsChannel,
             CancellationToken cancellationToken,
             ISqServerSentEventParser sqServerSentEventParser)
         {
-            this.serverStream = serverStream;
+            this.networkStreamReader = networkStreamReader;
             this.cancellationToken = cancellationToken;
             this.sqEventsChannel = sqEventsChannel;
             this.sqServerSentEventParser = sqServerSentEventParser;
         }
 
-        public Task BeginListening()
+        public async Task BeginListening()
         {
-            // todo: next PR
-            throw new NotImplementedException();
+            var eventLines = new List<string>();
+
+            while (!networkStreamReader.EndOfStream && !cancellationToken.IsCancellationRequested)
+            {
+                var line = await networkStreamReader.ReadLineAsync();
+                var isEventEnd = string.IsNullOrEmpty(line);
+
+                if (isEventEnd)
+                {
+                    var parsedEvent = sqServerSentEventParser.Parse(eventLines);
```

Comment:
Will the parser ever return `null`? (e.g. an unrecognised event)

Could the parser throw? If so, that is something we could recover from here since the stream reader is still in a known state i.e. we've read a series of related lines that comprise a logical event. If there was a problem parsing the event we could just ignore it an continue to read the next event.

Reply:
The parser should not throw but you're right it can return null, I'll fix